widget/editor.go

```diff
@@ -69,15 +69,25 @@ type Editor struct {
 		// position when moving between lines.
 		xoff fixed.Int26_6
 
-		// line is the caret line position as an index into lines.
-		line int
-		// col is the caret column measured in runes.
-		col int
-		// (x, y) are the caret coordinates.
-		x fixed.Int26_6
-		y int
+		// pos.lineCol.Y is the caret line position as an index into lines.
+		// pos.lineCol.X is the caret column measured in runes.
+		// pos.(x, y) are the caret coordinates.
+		// pos.ofs is the caret offset into the text, from rr.caret
+		//
+		// Use Editor.copyCaret to copy this field, e.g. in selection-related
+		// code. caret.pos.ofs is not kept up-to-date with rr.caret; copyCaret()
```

Comment:
Similar to the previous question about redundant combinedPos values, can rr.caret always be derived from pos or selStart? If so, can it be deleted?

Reply:
So far, I don't see a good way to do either of those things.

editorBuffer uses caret in 16 places.  Editor uses rr.caret in 40 places.  It seems like editorBuffer needs to keep caret, and, at most, Editor.caret.pos.ofs should somehow be made an alias to it.  I could make it a pointer to rr.caret, but that makes other uses of combinedPos annoying to deal with.  And actually makes copying caret.pos just as problematic, since I'd need to allocate a new pointer, so I'd certainly write a function to do it.  So that doesn't really help.

I could pull Editor.caret.pos back out into its component pieces again, so pos.ofs didn't even exist.  That would make using offsetToScreenPos a little more awkward, since I'd need code to deal with Editor.caret and Editor.selStart differently.  On the other hand I only call that twice, and I could write a helper function to encapsulate that operation.

I could make sure I update caret.pos.ofs every time rr.caret changes, and I think I tried that, but it seemed likely that later contributors wouldn't realize they needed to do the same and pos.ofs and rr.caret could easily get out of sync.

Thoughts / suggestions?

Reply:
Because Editor needs the richest caret information, I believe the best way forward is to eliminate editBuffer.caret. It's correct that there are 16 references to caret in editBuffer, but a closer look reveals that only 2 methods need caret: moveGap and prepend (deleteRunes only update caret). Given that, adding a position argument to moveGap and prepend doesn't seem unreasonable.